File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/NamedRangeMatcher.java
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern;

import java.util.function.Consumer;
import java.util.function.Function;

public record NamedRangeMatcher(String fromName, String toName) implements NamedMatcher<RangeMatcher> {
    @Override
    public void getNameReferences(Consumer<String> consumer) {
        consumer.accept(fromName);
        consumer.accept(toName);
    }

    @Override
    public RangeMatcher resolveNames(Function<String, String> resolver) {
        final int from = toCodePoint(resolver.apply(fromName));
        final int to = toCodePoint(resolver.apply(toName));
        return from <= to ? new RangeMatcher(from, to) : new RangeMatcher(to, from, true);
    }

    private static int toCodePoint(String s) {
        return s.length() > 0 ? s.codePointAt(0) : 0;
    }
}

File: fabric_mod_server/src/main/kotlin/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/NamedRangeMatcher.kt
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern

data class NamedRangeMatcher(val fromName: String, val toName: String) : NamedMatcher<RangeMatcher> {
    override fun getNameReferences() = sequence {
        yield(fromName)
        yield(toName)
    }

    override fun resolveNames(resolve: (String) -> String): RangeMatcher {
        val from = toCodePoint(resolve(fromName))
        val to = toCodePoint(resolve(toName))
        return if (from <= to) RangeMatcher(from, to) else RangeMatcher(to, from, true)
    }

    companion object {
        private fun toCodePoint(s: String): Int {
            return if (s.isNotEmpty()) s.codePointAt(0) else 0
        }
    }
}